src/System.Management.Automation/engine/NativeCommandProcessor.cs

```diff
@@ -528,90 +541,431 @@ internal override void Complete()
                             }
                         }
                     }
+                }
+                catch (Exception)
+                {
+                    StopProcessing();
+                    throw;
+                }
+
+                if (_isRunningInBackground == false)
+                {
+                    InitOutputQueue();
+                }
+            }
+            catch (Win32Exception e)
+            {
+                exceptionToRethrow = e;
+
+            } // try
+            catch (PipelineStoppedException)
+            {
+                // If we're stopping the process, just rethrow this exception...
+                throw;
+            }
+            catch (Exception e)
+            {
+                CommandProcessorBase.CheckForSevereException(e);
+
+                exceptionToRethrow = e;
+            }
+
+            // An exception was thrown while attempting to run the program
+            // so wrap and rethrow it here...
+            if (exceptionToRethrow != null)
+            {
+                // It's a system exception so wrap it in one of ours and re-throw.
+                string message = StringUtil.Format(ParserStrings.ProgramFailedToExecute,
+                    this.NativeCommandName, exceptionToRethrow.Message,
+                    this.Command.MyInvocation.PositionMessage);
+                ApplicationFailedException appFailedException = new ApplicationFailedException(message, exceptionToRethrow);
+
+                // There is no need to set this exception here since this exception will eventually be caught by pipeline processor.
+                // this.commandRuntime.PipelineProcessor.ExecutionFailed = true;
+
+                throw appFailedException;
+            }
+        }
+
+        private class ProcessOutputHandler
```

Comment:
Can you please move this class outside of `NativeCommandProcessor`? My arguments:
1. Nested class usually worsens readability of the code.
2. It's essentially the counterpart of `ProcessInputWriter`, and thus I think it should be a non-nested internal class similar as `ProcessInputWriter`